src/Raven.Server/Documents/Replication/OutgoingDocumentReplication.cs

```diff
@@ -1,43 +1,46 @@
 ﻿using System;
 using System.Linq;
 using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
 using Raven.Abstractions.Logging;
 using Raven.Abstractions.Util;
 using Raven.Server.ReplicationUtil;
 using Raven.Server.ServerWide.Context;
 
 namespace Raven.Server.Documents.Replication
 {
-    public class OutgoingDocumentReplication
+    public class OutgoingDocumentReplication : IDisposable
     {
         private const string SystemDocumentPrefix = "Raven/";
         private readonly DocumentDatabase _database;
         protected long _lastSentEtag;
-        private bool _shouldWaitForChanges;
+        private bool _hasMoreDocumentsToSend;
         private readonly DocumentReplicationTransport _transport;
         private readonly ILog _log = LogManager.GetLogger(typeof (OutgoingDocumentReplication));
 
+        //private readonly SemaphoreSlim _replicationSemaphore = new SemaphoreSlim(0,1);
+
         public OutgoingDocumentReplication(
-            DocumentDatabase database, 
-            long lastSentEtag, 
+            DocumentDatabase database,             
             DocumentReplicationTransport transport)
         {
-            _database = database;
-            _lastSentEtag = lastSentEtag;
+            _database = database;           
             _transport = transport;
+            _lastSentEtag = -1;
         }
 
-        public bool ShouldWaitForChanges => _shouldWaitForChanges;
+        public bool HasMoreDocumentsToSend => _hasMoreDocumentsToSend;
 
         public void ExecuteReplicationOnce()
         {
+            AsyncHelpers.RunSync(() => _transport.EnsureConnectionAsync());
```

Comment:
You don't get to use this kind of code, this is client side, and this is on the server side.